src/Azure/Orleans.Persistence.AzureStorage/Providers/Storage/AzureBlobStorageOptions.cs

```diff
@@ -1,35 +1,80 @@
         /// </summary>
+        /// <remarks>
+        /// If this property contains a shared access signature, then no other credential properties are required.
+        /// Otherwise, the presence of any other credential property will take precedence over this.
+        /// </remarks>
         public Uri ServiceUri { get; set; }
 
         /// <summary>
-        /// Use AAD to access the storage account
+        /// Token credentials, to be used in conjunction with <see cref="ServiceUri"/>.
         /// </summary>
+        /// <remarks>
+        /// This property takes precedence over specifying only <see cref="ServiceUri"/> and over <see cref="ConnectionString"/>, <see cref="AzureSasCredential"/>, and <see cref="SharedKeyCredential"/>.
+        /// This property is superseded by <see cref="CreateClient"/>.
+        /// </remarks>
         public TokenCredential TokenCredential { get; set; }
 
         /// <summary>
-        /// Container name where grain stage is stored
+        /// Azure SAS credentials, to be used in conjunction with <see cref="ServiceUri"/>.
         /// </summary>
-        public string ContainerName { get; set; } = DEFAULT_CONTAINER_NAME;
-        public const string DEFAULT_CONTAINER_NAME = "grainstate";
+        /// <remarks>
+        /// This property takes precedence over specifying only <see cref="ServiceUri"/> and over <see cref="ConnectionString"/> and <see cref="SharedKeyCredential"/>.
+        /// This property is superseded by <see cref="CreateClient"/> and <see cref="TokenCredential"/>.
+        /// </remarks>
+        public AzureSasCredential AzureSasCredential { get; set; }
+
+        /// <summary>
+        /// Options to be used when configuring the blob storage client, or <see langword="null"/> to use the default options.
+        /// </summary>
+        public BlobClientOptions ClientOptions { get; set; }
+
+        /// <summary>
+        /// Shared key credentials, to be used in conjunction with <see cref="ServiceUri"/>.
+        /// </summary>
+        /// <remarks>
+        /// This property takes precedence over specifying only <see cref="ServiceUri"/> and over <see cref="ConnectionString"/>.
+        /// This property is superseded by <see cref="CreateClient"/>, <see cref="TokenCredential"/>, and <see cref="AzureSasCredential"/>.
+        /// </remarks>
+        public StorageSharedKeyCredential SharedKeyCredential { get; set; }
+
+        /// <summary>
+        /// The optional delegate used to create a <see cref="BlobServiceClient"/> instance.
+        /// </summary>
+        /// <remarks>
+        /// This property, if not <see langword="null"/>, takes precedence over <see cref="ConnectionString"/>, <see cref="SharedKeyCredential"/>, <see cref="AzureSasCredential"/>, <see cref="TokenCredential"/>, <see cref="ClientOptions"/>, and <see cref="ServiceUri"/>,
+        /// </remarks>
+        public Func<Task<BlobServiceClient>> CreateClient { get; set; }
```

Comment:
Passing the service provider could be useful here? Maybe also the provider name?

Reply:
The developer is able to capture that from the context when they configure the options:

``` C#
.UseOrleans((ctc, builder) => 
{
  builder.AddAzureBlobGrainStorage("name", optionsBuilder =>
    optionsBuilder.Configure((AzureBlobStorageOptions options, MyDependency dep1, ...) => ... ));
})
```

Do you think we should change the signature anyway?

Reply:
No you are right, it should be sufficient